[BOLT][BTI] Add MCPlusBuilder::insertBTI

bgergely0 opened the pull request:
This function contains most of the logic for BTI:
- it takes the BasicBlock and the instruction used to jump to it.
- then it checks if the first non-pseudo instruction is a sufficient
landing pad for the used call.
- if not, it generates the correct BTI instruction.

Also introduce the isCallCoveredByBTI helper to simplify the logic.

bgergely0 committed:
[BOLT][BTI] Add MCPlusBuilder::addBTItoBBStart

This function contains most of the logic for BTI:
- it takes the BasicBlock and the instruction used to jump to it.
- then it checks if the first non-pseudo instruction is a sufficient
landing pad for the used call.
- if not, it generates the correct BTI instruction.

Also introduce the isBTIVariantCoveringCall helper to simplify the logic.

peterwaller-arm commented on bolt/include/bolt/Core/MCPlusBuilder.h:
```diff
@@ -1888,6 +1888,19 @@ class MCPlusBuilder {
+
+  /// Adds a BTI landing pad to the start of the BB, that matches the indirect
+  /// call/jump inst.
+  virtual void addBTItoBBStart(BinaryBasicBlock &BB, MCInst &Call) const {
```
Nit: should it be addBTIToBBStart?

If we're entertaining a rename, how about `insert` rather than add - since normally I think of putting something at the beginning of a list as 'insertion' rather than 'addition'. `insertBTI` would be a clear verb for a function name (and 'ToBBStart would be a bit redundant since presumably we won't expect to insert a BTI anywhere else?)

peterwaller-arm replied:
ensureBTITarget?
makeBTITarget?

On the other hand `insertBTI` I think should be reasonably clear. If it begins with `if (alreadyBTITarget) return;` that shouldn't be too surprising to a developer given the purpose of BTI. At least I can't think of a good reason to need to double them up.

bgergely0 committed:
[BOLT][BTI] rename addBTItoBBStart to insertBTI

bgergely0 commented on bolt/lib/Target/AArch64/AArch64MCPlusBuilder.cpp:
```diff
@@ -2781,6 +2781,81 @@ class AArch64MCPlusBuilder : public MCPlusBuilder {
+
+  void addBTItoBBStart(BinaryBasicBlock &BB, MCInst &Call) const override {
+    auto II = BB.getFirstNonPseudo();
+    if (II != BB.end()) {
```
I think it's best to add an assert here, and check if the BB is empty on the caller's side.

bgergely0 replied:
added the assertion, plus an MCPlusBuilder unittest for it.

bgergely0 committed:
[BOLT] Add assert wehn calling insertBTI on empty BBs

BOLT may generate empty BBs, e.g. around function splitting, to hold
temporary labels. If they are the target of a new indirect branch, the
BTI should be inserted into the first "real" BasicBlock.

peterwaller-arm reviewed (commented):
The assertion needs fixing before this can land, but otherwise LGTM - will give approval once fixed. Also made some other minor suggestions.

bgergely0 committed:
[BOLT] allow adding BTI to empty BBs

bgergely0 commented on bolt/lib/Target/AArch64/AArch64MCPlusBuilder.cpp:
```diff
@@ -2838,18 +2838,17 @@ class AArch64MCPlusBuilder : public MCPlusBuilder {
+    if (!Empty) {
       if (isBTIVariantCoveringCall(Call, *II))
         return;
+    }
```
ahh formatter caught it as well....

bgergely0 committed:
[BOLT] Rename function to isCallCoveredByBTI

- this way the order of the parameters matches the order in the name of
  the function
